Refactor the Tian Approximation in the Reactive Fluid Transport Model

danieldouglas92 opened the pull request:
I was looking through the Reactive fluid transport model code and saw how the katz 2003 model is in it's own reaction model. I think I recall at the hackathon that the ultimate goal for this material model was to expand it by creating new reaction models, so I went and refactored the tian approximation into it's own reaction.

gassmoeller commented on source/material_model/reaction_model/tian2019_solubility.cc:
```diff
@@ -0,0 +1,212 @@
+
+      template <int dim>
+      Tian2019Solubility<dim>::Tian2019Solubility()
+        = default;
```
If you make the constructor the default constructor then you usually do not need to implement it at all. Just remove it from here and from the header file.

gassmoeller commented on include/aspect/material_model/reaction_model/tian2019_solubility.h:
```diff
@@ -0,0 +1,159 @@
+          // constructor
+          Tian2019Solubility();
+
+          /**
+           * Declare the parameters this function takes through input files.
+           */
+          static
+          void
+          declare_parameters (ParameterHandler &prm);
+
+          /**
+           * Read the parameters from the parameter file.
+           */
+          void
+          parse_parameters (ParameterHandler &prm);
```
Please put these functions at the end of the list of public functions here and in the source file. This is just a convention we (try to) follow, but it makes navigating the source files easier.
A typical class structure in ASPECT looks like this (with many functions being optional):

```

class ClassName
{
  constructor();

  initialize();

  all_other_minor_functions();

  main_function(in this case melt_fraction);

  declare_parameters();

  parse_parameters();

  private:

  ... all member variables and private functions
}
```

gassmoeller commented on include/aspect/material_model/reaction_model/tian2019_solubility.h:
```diff
@@ -0,0 +1,159 @@
+
+
+          /**
+           * Compute the free fluid fraction that can be present in the material based on the
```
`can be present` or `is present`? I think you mean `is present` because `can be present` would be 100%?

gassmoeller commented on include/aspect/material_model/reaction_model/tian2019_solubility.h:
```diff
@@ -0,0 +1,159 @@
+          /**
+           * Compute the free fluid fraction that can be present in the material based on the
+           * fluid content of the material and the fluid solubility for the given input conditions.
+           * @p in and @p melt_fractions need to have the same size.
```
there is no `melt_fractions` variable

gassmoeller commented on include/aspect/material_model/reaction_model/tian2019_solubility.h:
```diff
@@ -0,0 +1,159 @@
+           * @p in and @p melt_fractions need to have the same size.
+           *
+           * @param in Object that contains the current conditions.
+           * @param melt_fractions Vector of doubles that is filled with the
+           * allowable free fluid fraction for each given input conditions.
```
Update this to the new interface, i.e. explain what `porosity_idx` and `q` are.

gassmoeller commented on include/aspect/material_model/reaction_model/tian2019_solubility.h:
```diff
@@ -0,0 +1,159 @@
+           * pressure and temperature conditions. This is used to determine
+           * how free water interacts with the solid phase.
+           * @param in Object that contains the current conditions.
+           * @param q unsigned int from 0-3 indexing which rock phase the equilbrium
+           * bound water content is being calculated for
```
Oh, is that really the meaning of `q`? We usually use `q` as an index for the quadrature point to index into the vectors that are contained inside `in`. Please clarify and if you use this variable for something else, maybe find a different name unless it has to be `q` for other reasons.

gassmoeller commented on include/aspect/material_model/reaction_model/tian2019_solubility.h:
```diff
@@ -0,0 +1,159 @@
+                                                                   unsigned int q) const;
+
+        private:
+          /**
+           * Parameters for the solubility model of Tian et al., 2019.
+           */
```
I think this comment here is redundant now. All of these parameters are from Tian et al since this is the Tian et al plugin.

gassmoeller commented on include/aspect/material_model/reaction_model/tian2019_solubility.h:
```diff
@@ -0,0 +1,159 @@
+           * csat is the saturated mass fraction of water in the solid, and Td is the
+           * onset temperature of devolatilization for water.
+           */
+          std::vector<double> LR_peridotite_poly_coeffs {-19.0609, 168.983, -630.032, 1281.84, -1543.14, 1111.88, -459.142, 95.4143, 1.97246};
+          std::vector<double> csat_peridotite_poly_coeffs {0.00115628, 2.42179};
+          std::vector<double> Td_peridotite_poly_coeffs {-15.4627, 94.9716, 636.603};
+
+          std::vector<double> LR_gabbro_poly_coeffs {-1.81745, 7.67198, -10.8507, 5.09329, 8.14519};
+          std::vector<double> csat_gabbro_poly_coeffs {-0.0176673, 0.0893044, 1.52732};
+          std::vector<double> Td_gabbro_poly_coeffs {-1.72277, 20.5898, 637.517};
+
+          std::vector<double> LR_MORB_poly_coeffs {-1.78177, 7.50871, -10.4840, 5.19725, 7.96365};
+          std::vector<double> csat_MORB_poly_coeffs {0.0102725, -0.115390, 0.324452, 1.41588};
+          std::vector<double> Td_MORB_poly_coeffs {-3.81280, 22.7809, 638.049};
+
+          std::vector<double> LR_sediment_poly_coeffs {-2.03283, 10.8186, -21.2119, 18.3351, -6.48711, 8.32459};
+          std::vector<double> csat_sediment_poly_coeffs {-0.150662, 0.301807, 1.01867};
+          std::vector<double> Td_sediment_poly_coeffs {2.83277, -24.7593, 85.9090, 524.898};
```
All of these numbers are fixed and will never change during the program runtime. You can tell this to the compiler, which will then optimize the code further and also prevent you from accidentally changing the values by changing their type from `std::vector<double>` to `constexpr std::array<double,N>` where N is the number of values in this array. Initialization and access to these arrays can remain unchanged.

danieldouglas92 replied:
I had tried to do this with arrays originally but I ultimately ran into issues because all of the arrays are different lengths. Because I ultimately end up storing each of these `poly_coeffs` arrays into a vector farther down (Lines 138-148). It seems like there might be a way forward with tuples, when I tried this the looping over each `poly_coeffs` array seemed a little strange and this type of approach doesn't seem like it's done anywhere else in the code. 

I think the only way to get around the way that I'm currently storing these constant values would be to alter the way the code within the `tian_equilibrium_bound_water_content` function. But because all of these arrays are not necessarily the same size, I'm not sure that there's a good way to do this without adding a bunch of if statements.

gassmoeller commented on include/aspect/material_model/reaction_model/tian2019_solubility.h:
```diff
@@ -0,0 +1,159 @@
+           * and 50 GPa for sediment. These cutoff pressures were determined by extending the pressure range in Tian et al. (2019)
+           * and observing where the maximum allowed water contents jump towards infinite values.
+           */
+          const std::vector<double> pressure_cutoffs {10, 26, 16, 50};
```
same here, make this a `constexpr std::array`

danieldouglas92 replied:
This one vector can be made into a constexpr because it doesn't get stored in another vector thankfully!

gassmoeller commented on source/material_model/reactive_fluid_transport.cc:
```diff
@@ -475,19 +362,14 @@ namespace aspect
-                                     "if there is a compositional field called peridotite."));
               fluid_solid_reaction_scheme = tian_approximation;
+
+              prm.enter_subsection("Tian 2019 model");
+              {
+                tian2019_model.initialize_simulator (this->get_simulator());
+                tian2019_model.parse_parameters(prm);
+              }
+              prm.leave_subsection();
```
So you put the parameters of the Tian model in their own subsection now. However, this means the parameter files of everyone using the reactive fluid flow model with the 3.0 version will need to be updated. Please include a changelog entry that mentions this. Additionally, you could extend our update scripts in `contrib/utilities/update_scripts/prm_files/` to automatically move this parameter. The closest operation that does something like this is probably [this function](https://github.com/geodynamics/aspect/blob/27bc5779e070c00e23f7fb951bc3a31bbe0164c8/contrib/utilities/update_scripts/prm_files/move_particles.py#L112). However, the update process looks a bit cryptic, I leave this up to you to decide if this is important enough.

danieldouglas92 replied:
I'll take a stab at this!

gassmoeller reviewed (commented):
Sorry for the delayed review. This looks generally good, I just have a few comments. Please add a changelog entry that mentions this plugin is now easier to include in material models.

danieldouglas92 commented:
@gassmoeller Thanks for the review, besides the point about creating std::arrays I've addressed all your points, created a python script to modify the input file and automatically create a new subsection for the Tian 2019 reaction model. This is now ready for another review when you get the chance!

gassmoeller reviewed (approved):
Very nice, thanks for the improvements and for trying the arrays, I didnt catch that they have different sizes.

gassmoeller commented:
/rebuild
